crates/pep508-rs/src/marker/tree.rs

```diff
@@ -808,141 +596,187 @@ impl MarkerTree {
+    /// never both evaluate to `true` in a given environment. However, this method may return
+    /// false negatives, i.e. it may not be able to detect that two markers are disjoint for
+    /// complex expressions.
+    pub fn is_disjoint(&self, tree: &MarkerTree) -> bool {
+        INTERNER.lock().and(self.0, tree.0).is_false()
     }
 
-    /// Remove the extras from a marker, returning `None` if the marker tree evaluates to `true`.
+    /// Returns the contents of this marker tree, if it contains at least one expression.
     ///
-    /// Any `extra` markers that are always `true` given the provided predicate will be removed.
-    /// Any `extra` markers that are always `false` given the provided predicate will be left
-    /// unchanged.
+    /// If the marker is `true`, this method will return `None`.
+    /// If the marker is `false`, the marker is represented as the normalized expression, `python_version < '0'`.
     ///
-    /// For example, if `is_extra('dev')` is true, given
-    /// `sys_platform == 'linux' and extra == 'dev'`, the marker will be simplified to
-    /// `sys_platform == 'linux'`.
-    pub fn simplify_extras_with(self, is_extra: impl Fn(&ExtraName) -> bool) -> Option<MarkerTree> {
-        // Because `simplify_extras_with_impl` is recursive, and we need to use
-        // our predicate in recursive calls, we need the predicate itself to
-        // have some indirection (or else we'd have to clone it). To avoid a
-        // recursive type at codegen time, we just introduce the indirection
-        // here, but keep the calling API ergonomic.
-        self.simplify_extras_with_impl(&is_extra)
+    /// The returned type implements [`Display`] and [`serde::Serialize`].
+    pub fn contents(&self) -> Option<MarkerTreeContents> {
+        if self.is_true() {
+            return None;
+        }
+
+        Some(MarkerTreeContents(self.clone()))
     }
 
-    fn simplify_extras_with_impl(
-        self,
-        is_extra: &impl Fn(&ExtraName) -> bool,
-    ) -> Option<MarkerTree> {
-        /// Returns `true` if the given expression is always `true` given the set of extras.
-        fn is_true(expression: &MarkerExpression, is_extra: impl Fn(&ExtraName) -> bool) -> bool {
-            match expression {
-                MarkerExpression::Extra {
-                    operator: ExtraOperator::Equal,
-                    name,
-                } => is_extra(name),
-                MarkerExpression::Extra {
-                    operator: ExtraOperator::NotEqual,
-                    name,
-                } => !is_extra(name),
-                _ => false,
-            }
+    /// Returns a simplified string representation of this marker, if it contains at least one
+    /// expression.
+    ///
+    /// If the marker is `true`, this method will return `None`.
+    /// If the marker is `false`, the marker is represented as the normalized expression, `python_version < '0'`.
```

Comment:
I think this was a nice API decision. Prevents one from emitting an "invalid" marker, but still makes it easy to print a debug repr.